include/tvm/relax/dataflow_analysis.h

```diff
@@ -0,0 +1,198 @@
+#define TVM_RELAX_DATAFLOW_ANALYSIS_H_
+
+#include <tvm/relax/analysis.h>
+#include <tvm/relax/expr.h>
+#include <tvm/runtime/object.h>
+
+#include <utility>
+
+namespace tvm {
+namespace relax {
+
+/*! \brief For dataflow analysis, we need to have a control flow graph.
+ *  We will organize this graphs by bindings, which allows analyses to
+ *  state their results for each binding in a SeqExpr.
+ *
+ *  There are a few cases that have to be handled:
+ *  1. A normal binding (most common)ICHECK
+ *  2. The condition expression in an If node (a "split" point)
+ *  3. A merge point (the variable to which an If node is bound: it is a "merge" between
+ *     the SeqExprs in the true and false branches)
+ *  4. The body expression in a SeqExpr (not actually bound)
+ */
+enum BindingNodeKind : int { kBinding = 0, kIfCond = 1, kIfMerge = 2, kSeqBody = 3 };
+
+class GraphBindingNode : public Object {
+ public:
+  /*! \brief The SeqExpr the binding resides in. */
+  SeqExpr seq;
+
+  /*! \brief The arguments to the binding. Only the first binding in the graph has arguments
+   * (i.e., the function arguments). */
+  Array<Var> args;
+
+  /*! \brief Index of the binding block in the SeqExpr where the binding is found.
+   *  Convention: We put the SeqExpr body at one block past the final block. */
+  size_t block_idx;
+
+  /*! \brief Index of the binding within the binding block corresponding to this binding.
+   *  Convention: Both the If condition and merge are mapped to the same index.
+   *  We use the kind to distinguish. */
+  size_t binding_idx;
+
+  /*! \brief The kind of binding this is. */
+  BindingNodeKind kind;
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("seq", &seq);
+    v->Visit("args", &args);
+    v->Visit("block_idx", &block_idx);
+    v->Visit("binding_idx", &binding_idx);
+    v->Visit("kind", &kind);
+  }
+
+  static constexpr const uint32_t _type_index = TypeIndex::kDynamic;
+  static constexpr const char* _type_key = "relax.analysis.GraphBinding";
+  TVM_DECLARE_BASE_OBJECT_INFO(GraphBindingNode, Object);
+};
+
+/*! \brief Representation of a binding in the control flow graph */
+class GraphBinding : public ObjectRef {
+ public:
+  /*!
+   * \brief Create a GraphBinding. See the docs on GraphBindingNode for further details.
+   *
+   * \param seq: The SeqExpr in which the binding resides.
+   * \param args: The arguments to the binding (only nonempty for the first binding:
+   *   these will be the function arguments)
+   * \param block_idx: The index of the BindingBlock in the SeqExpr
+   *   where the binding resides (for the return expression, use one past the final block).
+   * \param binding_idx: The index of the binding in the BindingBlock corresponding to the binding.
+   * \param kind: The kind of binding this is. (Used especially to distinguish If node conditions
+   * from the merge after the If)
+   */
+  TVM_DLL static GraphBinding Create(const SeqExpr& seq, const Array<Var>& args, size_t block_idx,
```

Comment:
I was following the pattern of many other ObjectRefs, but we could do it as a constructor instead.

Reply:
Makes sense.  The other cases I've seen tend to have multiple valid constructors with equivalent type signatures (e.g. `(Expr begin, Expr end)` vs `(Expr min, Expr extent)`), and the static functions are required to avoid ambiguity.